samples/nrf9160/modem_shell/src/startup_cmd/startup_cmd_ctrl.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <stdio.h>
+
+#include <zephyr/shell/shell.h>
+#include <zephyr/shell/shell_uart.h>
+
+#include "mosh_print.h"
+#include "startup_cmd_settings.h"
+
+extern struct k_work_q mosh_common_work_q;
+
+/* Work for running commands: */
+struct startup_cmd_worker_data {
+	struct k_work_delayable work;
+	uint8_t running_mem_slot;
+};
+static struct startup_cmd_worker_data startup_cmd_worker_data;
+
+static bool running;
+
+static void startup_cmd_worker(struct k_work *work_item)
+{
+	const struct shell *shell = shell_backend_uart_get_ptr();
+	char *shell_cmd_str;
+	int len = 0;
+	struct startup_cmd_worker_data *data_ptr =
+		CONTAINER_OF(work_item, struct startup_cmd_worker_data, work);
+
+	shell_cmd_str = startup_cmd_get(data_ptr->running_mem_slot);
+	if (shell_cmd_str == NULL) {
+		mosh_error("%s shell_cmd_str == NULL\n", __func__);
+		return;
+	}
+
+	len = strlen(shell_cmd_str);
+	if (len) {
+		mosh_print("Starting to execute sett_cmd %d \"%s\"...",
+			data_ptr->running_mem_slot, shell_cmd_str);
+		shell_execute_cmd(shell, shell_cmd_str);
+	}
+
+	/* We are done for this one. Schedule next stored command. */
+	if (data_ptr->running_mem_slot < STARTUP_CMD_MAX_COUNT) {
+		k_work_init_delayable(&startup_cmd_worker_data.work, startup_cmd_worker);
```

Comment:
`k_work_init_delayable()` is called from three places. This is unnecessary, because a work item only needs to be initialized once, before it's used for the first time.  The API also says that the init function must only be called when the work item is idle, so technically calling it from `startup_cmd_worker()` is not supported (although it doesn't seem to cause issues here, I guess).

I would suggest calling it from `startup_cmd_ctrl_init()` instead. That should guarantee that the item is initialized before it's used for the first time and that init is not called when the work item is not idle, right?

Reply:
I was able to repro the assert that I was having previously:
ASSERTION FAIL [z_spin_lock_valid(l)] @ WEST_TOPDIR/zephyr/include/zephyr/spinlock.h:142
        Recursive spinlock

But with the correct initialization order, where work is initialized before settings, this is not happening and it is possible to have only one work initi as you suggested. Changes now done.